Add interface to collect RPC-based training metrics.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33153 Add interface to collect RPC-based training metrics.**

Adds the initial interface `RpcMetricsHandler.h` for tracking metrics during distributed model parallel/RPC-based training. Metric Handlers should implement this interface in order to log metrics and stream them to some metrics storage.

Differential Revision: [D19615364](https://our.internmc.facebook.com/intern/diff/D19615364/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19615364/)!

rohan-varma committed:
[rpc][metrics] add initial metric handler classes.

These are the initial classes and unit tests for RPC metrics. They contain:
`RpcMetrics.h` - the abstract interface from which metrics handler will inherit (i.e. the ServiceDataMetricsHandler will inherit from this)
`ConsoleMetricsHandler` - a dummy metrics handler that keeps track of aggregate and individual metrics. Not used in prod but will help with debugging and can also be a simple OSS interface.

If these interfaces look okay, we can export them to OSS in a later diff and remove the interfaces from internal-FB code.

Per the
quip design doc https://fb.quip.com/s0u3AffEJzbO, the next steps will be to implement logging to ODS via fbData
and explore using Logger to log to both ODS and Scuba. We will then also
integrate this with ThriftRpcAgent and ensure we are capturing a desired set of
metrics.

Differential Revision: [D19615364](https://our.internmc.facebook.com/intern/diff/D19615364/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19615364/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Add interface to collect RPC-based training metrics."


Adds the initial interface `RpcMetricsHandler.h` for tracking metrics during distributed model parallel/RPC-based training. Metric Handlers should implement this interface in order to log metrics and stream them to some metrics storage.

Differential Revision: [D19615364](https://our.internmc.facebook.com/intern/diff/D19615364/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19615364/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/metrics/RpcMetricsHandler.h:
```diff
@@ -0,0 +1,27 @@
+// from multiple threads without the user needing to coordinate serialization.
+// The class provides two main functions - accumulateMetric which can be used
+// for statistical aggregations and setMetric for setting/modifying single
+// values.
```
We should mention that implementations of this class should provide these APIs in a thread safe manner. Since this class is just an interface, it can't provide any Thread safety guarantees.

pritamdamania87 commented on torch/csrc/distributed/rpc/metrics/RpcMetricsHandler.h:
```diff
@@ -0,0 +1,27 @@
+class RpcMetricsHandler {
+ public:
+  // Log value corresponding to metric name given by name to the time-series
+  // metric stream.
```
This seems like implementation detail. The doc should say something like this "Accumulates the metric value specified for purposes of computing average/percentiles over time".

pritamdamania87 commented on torch/csrc/distributed/rpc/metrics/RpcMetricsHandler.h:
```diff
@@ -0,0 +1,27 @@
+  // metric stream.
+  virtual void accumulateMetric(const std::string& name, double value) = 0;
+  // Increment a count for the metric given by the name.
+  virtual void bumpMetric(const std::string& name) = 0;
```
nit: incrementMetric is a better name.

rohan-varma committed:
Update on "Add interface to collect RPC-based training metrics."


Adds the initial interface `RpcMetricsHandler.h` for tracking metrics during distributed model parallel/RPC-based training. Metric Handlers should implement this interface in order to log metrics and stream them to some metrics storage.

Differential Revision: [D19615364](https://our.internmc.facebook.com/intern/diff/D19615364/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19615364/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Add interface to collect RPC-based training metrics."


Adds the initial interface `RpcMetricsHandler.h` for tracking metrics during distributed model parallel/RPC-based training. Metric Handlers should implement this interface in order to log metrics and stream them to some metrics storage.

Differential Revision: [D19615364](https://our.internmc.facebook.com/intern/diff/D19615364/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19615364/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Add interface to collect RPC-based training metrics."


Adds the initial interface `RpcMetricsHandler.h` for tracking metrics during distributed model parallel/RPC-based training. Metric Handlers should implement this interface in order to log metrics and stream them to some metrics storage.

Differential Revision: [D19615364](https://our.internmc.facebook.com/intern/diff/D19615364/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19615364/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Add interface to collect RPC-based training metrics."


Adds the initial interface `RpcMetricsHandler.h` for tracking metrics during distributed model parallel/RPC-based training. Metric Handlers should implement this interface in order to log metrics and stream them to some metrics storage.

Differential Revision: [D19615364](https://our.internmc.facebook.com/intern/diff/D19615364/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19615364/)!

[ghstack-poisoned]
